cluster-autoscaler/processors/scaledowncandidates/scale_down_candidates_delay_processor.go

```diff
@@ -0,0 +1,104 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaledowncandidates
+
+import (
+	"reflect"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	"k8s.io/autoscaler/cluster-autoscaler/clusterstate"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+)
+
+// ScaleDownCandidatesDelayProcessor is a processor to filter out
+// nodes according to scale down delay per nodegroup
+type ScaleDownCandidatesDelayProcessor struct {
+	clusterStateRegistry *clusterstate.ClusterStateRegistry
+}
+
+// GetPodDestinationCandidates returns nodes as is no processing is required here
+func (p *ScaleDownCandidatesDelayProcessor) GetPodDestinationCandidates(ctx *context.AutoscalingContext,
+	nodes []*apiv1.Node) ([]*apiv1.Node, errors.AutoscalerError) {
+	return nodes, nil
+}
+
+// GetScaleDownCandidates returns filter nodes based on if scale down is enabled or disabled per nodegroup.
+func (p *ScaleDownCandidatesDelayProcessor) GetScaleDownCandidates(ctx *context.AutoscalingContext,
+	nodes []*apiv1.Node) ([]*apiv1.Node, errors.AutoscalerError) {
+	result := []*apiv1.Node{}
+
+	for _, node := range nodes {
+		nodeGroup, err := ctx.CloudProvider.NodeGroupForNode(node)
+		if err != nil {
+			klog.Warningf("Error while checking node group for %s: %v", node.Name, err)
+			continue
+		}
+		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+			klog.V(4).Infof("Node %s should not be processed by cluster autoscaler (no node group config)", node.Name)
+			continue
+		}
+
+		lastScaledUps := p.clusterStateRegistry.GetScaleUpRequests()
+		lastScaleDowns := p.clusterStateRegistry.GetScaleDownRequests()
+		lastScaleUpFailures := p.clusterStateRegistry.GetScaleUpFailures()
+		currentTime := time.Now()
+
+		if lastScaledUps[nodeGroup.Id()] != nil &&
+			lastScaledUps[nodeGroup.Id()].ExpectedAddTime.After(currentTime) {
+			klog.V(4).Infof("Skipping scale down on node group %s because it was scaled up recently at %v", nodeGroup.Id(), lastScaledUps[nodeGroup.Id()].Time)
+			continue
+		}
+
+		if lastScaleDowns[nodeGroup.Id()] != nil &&
+			lastScaleDowns[nodeGroup.Id()].ExpectedDeleteTime.After(currentTime) {
+			klog.V(4).Infof("Skipping scale down on node group %s because it was scaled down recently at %v", nodeGroup.Id(), lastScaleDowns[nodeGroup.Id()].Time)
+			continue
+		}
+
+		foundRecentFailure := false
+		if len(lastScaleUpFailures[nodeGroup.Id()]) > 0 {
+			for _, failure := range lastScaleUpFailures[nodeGroup.Id()] {
+				if failure.Time.Add(ctx.ScaleDownDelayAfterFailure).After(currentTime) {
+					klog.V(4).Infof("Skipping scale down on node group %s because it failed to scale up recently at %v", nodeGroup.Id(), failure.Time)
```

Comment:
Too many fors and ifs here. Better ideas are welcome :pray:

Reply:
Thank you for the review comment. Not sure if I understand it right:
Here's what I think I am doing right now:
```go
for _, node := range nodes {
        ...
		foundRecentFailure := false
		if len(lastScaleUpFailures[nodeGroup.Id()]) > 0 {
			for _, failure := range lastScaleUpFailures[nodeGroup.Id()] {
				if failure.Time.Add(ctx.ScaleDownDelayAfterFailure).After(currentTime) {
					klog.V(4).Infof("Skipping scale down on node group %s because it failed to scale up recently at %v", nodeGroup.Id(), failure.Time)
					foundRecentFailure = true
					break
				}
			}
		}
		if foundRecentFailure {
			continue
		}
        result = append(result, node) // nodes to scale down
}
```
I guess you are suggesting something like this:
```go
for _, node := range nodes {
        ...
		foundRecentFailure := false
		if len(lastScaleUpFailures[nodeGroup.Id()]) > 0 {
			for _, failure := range lastScaleUpFailures[nodeGroup.Id()] {
				if failure.Time.Add(ctx.ScaleDownDelayAfterFailure).Before(currentTime) {
					continue
				}
			}
		}

        result = append(result, node) // nodes to scale down
}
```
I see how this is cleaner. But this code doesn't stop going to next iteration if `failure.Time.Add(ctx.ScaleDownDelayAfterFailure).After(currentTime)` is `true`. Using `break` in the earlier code stops the code from going to next iteration and saves some processing if `failure.Time.Add(ctx.ScaleDownDelayAfterFailure).After(currentTime)` is `true`.

Reply:
Get. maybe we can move these codes to another function, and some variables can put outside the for loop, like this：

<img width="1920" alt="image" src="https://github.com/kubernetes/autoscaler/assets/5654791/dcf0f5d4-b9ba-493f-822a-c4559da52fd5">

Reply:
@qianlei90 thanks for the elaborating. That makes sense. I did think about putting the code into a separate function. I was hoping there would be other ways to reduce the number of `if`s and `for`s. Looks like abstracting it out into a function is probably the way to go (even though it doesn't reduce the total number of `for`s and `if`s, it makes things cleaner which I like).  

Once we get an OK from Daniel around finalizing the approach [here](https://github.com/kubernetes/autoscaler/pull/5729#discussion_r1200575895), I will abstract the code out into a separate function (based on your suggestion).

Reply:
Resolving this. Please unresolve if needed.